### PR TITLE
refactor: contract upgradeable

### PR DESCRIPTION
**Pull request by tnkshuuhei**
### Factory

| Network (chainId) | Address                                                                                                                          |
| ----------------- | -------------------------------------------------------------------------------------------------------------------------------- |
| 10                | [0x2300cb3e09733b6F7390328976CdD7878f068877](https://optimistic.etherscan.io/address/0x2300cb3e09733b6F7390328976CdD7878f068877) |
| 11155111          | [0xa57d81bDD038fCFEC275580b2d3F38dfD8B125dc](https://sepolia.etherscan.io/address/0xa57d81bDD038fCFEC275580b2d3F38dfD8B125dc)    |

### Yohaku (Proxy)

| Network (chainId) | Proxy Address                                                                                                                    | Implementation Address                                                                                                           |
| ----------------- | -------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
| 10                | [0x241b846142C5b06C904db55438248e2c98cA55d6](https://optimistic.etherscan.io/address/0x241b846142C5b06C904db55438248e2c98cA55d6) | [0x1F6385D8409C23AEB0641C9176F46c4C09520CD5](https://optimistic.etherscan.io/address/0x1F6385D8409C23AEB0641C9176F46c4C09520CD5) |
| 11155111          | [0x241b846142C5b06C904db55438248e2c98cA55d6](https://sepolia.etherscan.io/address/0x241b846142C5b06C904db55438248e2c98cA55d6)    | [0x1F6385D8409C23AEB0641C9176F46c4C09520CD5](https://sepolia.etherscan.io/address/0x1F6385D8409C23AEB0641C9176F46c4C09520CD5)    |

## Task
- [x] basic refactor
- [x] test upgrade
- [x] coverage
- [x] add documentation
- [x] add NatSpec

## Summary by CodeRabbit

- **New Features**
  - Added deployment scripts and configuration files for automated contract deployment and upgrades.
  - Introduced deterministic NFT deployment with address precomputation and prevention of duplicate deployments.
  - Enabled contract upgradeability and pausing features with enhanced role management.
  - Added support for specifying image URLs during NFT minting and improved token metadata handling.
  - Introduced versioning for contracts and a standard interface to expose version information.

- **Improvements**
  - Enhanced access control with role-based permissions and custom error handling across contracts.
  - Updated Solidity compiler version to 0.8.28 and improved code formatting and documentation.
  - Expanded and reorganized README for better clarity, navigation, and setup instructions.
  - Updated configuration files for improved formatting, optimization, and environment variable management.
  - Updated environment example file with clearer placeholders and added new environment variables.

- **Bug Fixes**
  - Improved error handling and revert conditions for batch operations and role checks.

- **Tests**
  - Significantly expanded test coverage for new features, role management, upgrade mechanics, and edge cases.

- **Chores**
  - Updated dependencies and submodules to newer versions.
  - Added configuration files for static analysis and improved `.gitignore`.

- **Style**
  - Reformatted code and improved readability in contracts and mock test files.

- **Documentation**
  - Added NatSpec and inline comments, and updated placeholder values in environment example files.

- **Refactor**
  - Reorganized contract structure, removed redundant scripts, and streamlined deployment logic.

### CODE REVIEW

File: .github/scripts/comment.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  await github.rest.issues[commentFn]({
+    owner: context.repo.owner,
+    repo: context.repo.repo,
+    body: comment,
+    ...(botComment
+      ? { comment_id: botComment.id }
+      : { issue_number: context.payload.number }),
+  });
+};
```
[the remark this answers is not in the document]
Comment (Naokiakazawa) [reply .github/scripts/comment.js:26]: このコードって何用のGitHub Actionsだっけ？

File: contracts/ContributionNFT.sol [hunk 2]
```diff
@@ -1,79 +1,73 @@
...
+        address owner;
+        string description;
+        string imageUrl;
     }
 
-    modifier onlyAdmin() {
-        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender), "Caller is not a admin");
-        _;
-    }
+    event Minted(address indexed to, address indexed account, bytes32 indexed attestationUID);
```
Comment (Naokiakazawa) [inline contracts/ContributionNFT.sol:43]: イベント引数にtokenIdを入れておいてください。

Comment (Naokiakazawa) [reply contracts/ContributionNFT.sol:43]: あ、でも、attestationUIDで特定できるんかこれ。

File: contracts/ContributionNFT.sol [hunk 3]
```diff
@@ -1,79 +1,73 @@
...
+    error ALREADY_HAVE_TOKEN(address owner);
+    error INVALID_MINTER(address minter);
+    error INVALID_ADMIN(address admin);
+    error LENGTH_MISMATCH(uint256 recipient, uint256 account);
 
-    modifier onlyMinter() {
-        require(hasRole(MINTER_ROLE, msg.sender), "Caller is not a minter");
-        _;
+    struct TokenData {
+        address owner;
```
Comment (Naokiakazawa) [inline contracts/ContributionNFT.sol:38]: 確認だけど、このフィールド、仮に転送された際でも、ミント時のオーナーだけが保持されるよね？

File: contracts/Yohaku.sol [hunk 4]
```diff
@@ -1,108 +1,200 @@
...
+    /// @notice mapping of token ID to TokenData struct
+    mapping(uint256 => TokenData) public tokenData;
+
+    /// @notice mapping of token ID to previous owners
+    mapping(uint256 => address[]) public previousOwners;
+
+    error ALREADY_HAVE_TOKEN(address owner);
 
     struct TokenData {
         address owner;
```
Comment (Naokiakazawa) [inline contracts/Yohaku.sol:47]: このフィールド、仮にYohakuが転送された時に更新されずに残る？

**Review by Naokiakazawa — COMMENTED** [review]
i made several comments.

File: .github/scripts/comment.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnkshuuhei) [reply .github/scripts/comment.js:26]: exactly

File: contracts/ContributionNFT.sol (hunk 2 above)
Comment (tnkshuuhei) [reply contracts/ContributionNFT.sol:43]: > イベント引数にtokenIdを入れておいてください。
> あ、でも、attestationUIDで特定できるんかこれ。

yohaku or ContributionNFTどちらのtokenIdのことを指していますか？
attestationUIDに入っているのはyohakuのtokenIdです
`"address TokenBoundAccount,address CurrentOwner,address TokenAddress,uint256 tokenId,uint8 Score,string Description",`
ContributionNFTに関しては、tokenIdのtransfer自体はOZが`Transfer`eventを発火させるので問題ないかと思います。

どちらかといえばこのイベントはattestationを取得する目的です。
https://docs.openzeppelin.com/contracts/5.x/api/token/erc721#IERC721-Transfer-address-address-uint256-

File: contracts/ContributionNFT.sol (hunk 3 above)
Comment (tnkshuuhei) [reply contracts/ContributionNFT.sol:38]: ContributionNFT.solに関してはそうなっていますね。
元々Yohaku.solのコードをもとにしている名残で残っていますが、TokenData.ownerは特に使っていないので消してもいいかもです

yohakuに関しては、Transferすると更新されます
https://github.com/yohakuxyz/yohaku-contract/blob/e73acc490a83314e40f52b1fb2c1ddb32e43cb32/contracts/Yohaku.sol#L184-L199

File: contracts/Yohaku.sol (hunk 4 above)
Comment (tnkshuuhei) [reply contracts/Yohaku.sol:47]: `previousOwners`ですか？更新されます。
https://github.com/yohakuxyz/yohaku-contract/blob/e73acc490a83314e40f52b1fb2c1ddb32e43cb32/contracts/Yohaku.sol#L184-L199

File: contracts/ContributionNFT.sol (hunk 3 above)
Comment (tnkshuuhei) [reply contracts/ContributionNFT.sol:38]: @Naokiakazawa 確認お願いします

File: contracts/ContributionNFT.sol (hunk 2 above)
Comment (Naokiakazawa) [reply contracts/ContributionNFT.sol:43]: > attestationUIDに入っているのはyohakuのtokenIdです

なるほど、理解しました。
attestationを取得する用のイベントなのでこれでOKです

**Review by Naokiakazawa — COMMENTED** [review]
全て確認できました。問題ありません！

**Review by Naokiakazawa — APPROVED** [review]
LGTM
